node/src/chain/acceptor.rs

```diff
@@ -604,7 +580,133 @@ impl<DB: database::DB, VM: vm::VMExecution, N: Network> Acceptor<N, DB, VM> {
             );
         }
 
-        Ok(label)
+        Ok(finalized)
+    }
+
+    /// Perform the rolling finality checks, updating the database with new
+    /// labels if required
+    ///
+    /// Returns
+    /// - Current accepted block label
+    /// - Previous last finalized state root
+    /// - List of the new finalized state root
+    fn rolling_finality<D: database::DB>(
+        &self,
+        pni: u8,
+        blk: &Block,
+        db: &D::P<'_>,
+    ) -> Result<(Label, Option<RollingFinalityResult>)> {
+        let confirmed_after = match pni {
+            0 => 1u64,
+            n => 2 * n as u64,
+        };
+        let block_label = if pni == 0 {
+            Label::Attested(confirmed_after)
+        } else {
+            return Ok((Label::Accepted(confirmed_after), None));
+        };
+        let mut finalized_blocks = BTreeMap::new();
+
+        let current_height = blk.header().height;
+        let mut labels = BTreeMap::new();
+
+        // Retrieve latest blocks up to the last finalized
+        let mut old_final_hash = None;
+        for height in (0..current_height).rev() {
+            let (hash, label) = db.fetch_block_label_by_height(height)?.ok_or(
+                anyhow!("Cannot find block label for height {height}"),
+            )?;
+            if let Label::Final(_) = label {
+                old_final_hash = Some(hash);
+                break;
+            }
+            labels.insert(height, (hash, label));
+        }
+        let old_final_hash =
+            old_final_hash.expect("Unable to find last finalized block hash");
+        let old_final_state_root = db
+            .fetch_block_header(&old_final_hash)?
+            .ok_or(anyhow!(
+                "Cannot get header for last finalized block hash {}",
+                to_str(&old_final_hash)
+            ))?
+            .0
+            .state_hash;
+
+        // Iterate from tip to last_final to set label::confirmed
+        let mut stable_count = 1;
```

Comment:
Please explain what we consider as `stable` here, and that it's set to `1` because we're already counting `blk`